geoutils/georaster.py

```diff
@@ -822,10 +913,179 @@ def format_value(value):
+
+        """
+        i, j = self.ds.index(x,y)
+
+        return i, j
+
+    def ij2xy(self,i,j,offset='center'):
+
+        """
+        Return x,y coordinates for a given row, column index pair.
+
+        :param i: row (i) index of pixel.
+        :type i: array-like
+        :param j: column (j) index of pixel.
+        :type j: array-like
+        :param offset: return coordinates as "corner" or "center" of pixel
+        :param offset: str
+
+        :returns x, y: x,y coordinates of i,j in reference system.
+        """
+
+        x,y = self.ds.xy(i,j,offset=offset)
+
+        return x, y
+
+    def outside_image(self, xi,yj, index=True):
+        """
+        #TODO: calculate matricially for all points instead of doing for only one?
+        Check whether a given point falls outside of the raster.
+
+        :param xi: Indices (or coordinates) of x direction to check.
+        :param yj: Indices (or coordinates) of y direction to check.
+        :param index: Interpret ij as raster indices (default is True). If False, assumes ij is coordinates.
+        :type xi: array-like
+        :type yj: array-like
+        :type index: bool
+
+        :returns is_outside: True if ij is outside of the image.
+        """
+        if not index:
+            xi,xj = self.xy2ij(xi,yj)
+
+        if np.any(np.array((xi,yj)) < 0):
+            return True
+        elif xi > self.ds.width or yj > self.ds.height:
+            return True
+        else:
+            return False
+
+    def interp_points(self,pts,nsize=1,mode='linear',band=1):
+
+        """
+        Interpolate raster values at a given point, or sets of points.
+
+       :param pts: Point(s) at which to interpolate raster value. If points fall outside of image,
+       value returned is nan.'
+       :param nsize: Number of neighboring points to include in the interpolation. Default is 1.
+       :param mode: One of 'linear', 'cubic', or 'quintic'. Determines what type of spline is
+           used to interpolate the raster value at each point. For more information, see
+           scipy.interpolate.interp2d. Default is linear.
+       :param band: Raster band to use
+       :type pts: array-like
+       :type nsize: int
+       :type mode: str
+       :type band: int
+
+       :returns rpts: Array of raster value(s) for the given points.
+       :rtype rpts: array-like
+       """
+        assert mode in ['mean', 'linear', 'cubic', 'quintic',
+                        'nearest'], "mode must be mean, linear, cubic, quintic or nearest."
+
+        rpts = []
+
+        #TODO: might need to check if coordinates are center or point in the metadata here...
+
+
+        xx, yy = self.xxyy(offset='center', grid=False)
+        #TODO: right now it's a loop... could add multiprocessing parallel loop outside,
+        # but such a method probably exists already within scipy/other interpolation packages?
+        for pt in pts:
+            i,j = self.xy2ij(pt[0],pt[1])
```

Comment:
Here your i,j values are integers (by default with rio's ds.index), so I don't see the point of interpolating...
You should modify xy2ij to have an option to return the non-rounded indexes.

Reply:
Interpolation at the exact coordinates is done further down:
`                        zint = griddata((X.flatten(), Y.flatten()), z.flatten(), list(pt), method=mode)[0]
`
Here we just find the center of the window centered on the coordinates, to create smaller arrays passed to griddata (which is why  the process resembles quite a bit the "value_at_coords" method).

Reply:
Ok I understand. In this case, why not use value_at_coords to extract the pixels needed for interpolation? Did you notice the window option?